proposals/lifetime-variance.md

```diff
@@ -0,0 +1,141 @@
+
+# Lifetime variance
+
+This doc is meant as an addendum to https://github.com/dotnet/csharplang/blob/main/proposals/csharp-11.0/low-level-struct-improvements.md.
+
+The primary purpose is to more formally ground the variance of lifetime parameters and the corresponding correctness rules.
+
+In the above doc we treat ref variables as their own class of variables. However, by using a struct syntax we can unify the treatment of ref structs and by-ref variables.
+
+First, we will define a primitive type to represent a by-ref variable:
+
+```csharp
+ref struct ByRef<$a, T>
+{
+    public T Value { get; set; }
+}
+```
+
+This is the syntactic equivalent of `ref<$a> T`. One problem with this formulation is that C# currently defines all struct type parameters as invariant. However, this is not necessary. Structs are merely aggregations of their field data. That is, simple use of any struct type parameter `T` as a field type should not carry any additional variance restrictions. We do not currently have a syntax in C# to represent bivariance, but we can invent one: `inout`.
+
+With this syntax we can alter the definition to
+
+```csharp
+ref struct ByRef<inout $a, T>
+{
+    public T Value { get; set; }
+}
+```
+
+Note that `T` still remains invariant. This is for two reasons. First, because the intent of this doc is only to change struct variance for lifetimes, not for non-lifetime type parameters. Second, `ByRef` is a very special type of struct: it contains a pointer. Variance restrictions are due to a combination of two factors: mutability and aliasing. When a variable may be aliased but only allows reading, it is covariant. When it may be aliased but only allows writing, it is invariant. When it may be aliased and it allows both reading and writing, it is invariant. When it does not allow aliasing, it is bivariant, regardless of access. Because `T` exists behind a read/write pointer, it is invariant.
```

Comment:
"when it is aliased but only allows writing, it is invariant".  Should that be contravariant?  If not, can you clarify why?  Thanks!